src/Raven.Server/Documents/Sharding/Background/ShardedPeriodicDocumentsMigrator.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Threading.Tasks;
+using Raven.Server.Background;
+using Raven.Server.ServerWide.Commands.Sharding;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Logging;
+
+namespace Raven.Server.Documents.Sharding.Background
+{
+    public class ShardedPeriodicDocumentsMigrator : BackgroundWorkBase
+    {
+        private readonly ShardedDocumentDatabase _database;
+
+        public ShardedPeriodicDocumentsMigrator(ShardedDocumentDatabase database) : base(database.ShardedDatabaseName, database.ShardedDocumentsStorage.DocumentDatabase.DatabaseShutdown)
+        {
+            _database = database;
+        }
+
+        protected override async Task DoWork()
+        {
+            await WaitOrThrowOperationCanceled(_database.Configuration.Sharding.PeriodicDocumentsMigrationInterval.AsTimeSpan);
+
+            if (_database.ServerStore.Sharding.HasActiveMigrations(_database.ShardedDatabaseName))
+                return;
+
+            await ExecuteMoveDocuments();
+        }
+
+        internal async Task ExecuteMoveDocuments()
+        {
+            try
+            {
+                int bucket = -1;
+                int moveToShard = -1;
+                bool found = false;
+                using (_database.ShardedDocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                using (context.OpenReadTransaction())
+                {
+                    var configuration = _database.ShardingConfiguration;
+                    for (var index = 0; index < configuration.BucketRanges.Count; index++)
+                    {
+                        var range = configuration.BucketRanges[index];
+                        if (range.ShardNumber == _database.ShardNumber)
+                            continue;
+
+                        var start = range.BucketRangeStart;
+                        var end = index == configuration.BucketRanges.Count - 1
+                            ? int.MaxValue
+                            : configuration.BucketRanges[index + 1].BucketRangeStart - 1;
+
+                        var bucketStatistics = ShardedDocumentsStorage.GetBucketStatistics(context, start, end);
+
+                        foreach (var bucketStats in bucketStatistics)
+                        {
+                            if (bucketStats.NumberOfDocuments == 0)
+                                continue;
+
+                            bucket = bucketStats.Bucket;
+                            moveToShard = range.ShardNumber;
+                            found = true;
+                            break;
+                        }
+
+                        if (found)
+                            break;
+                    }
+                }
+
+                if (found)
+                    await MoveDocumentsToShard(bucket, moveToShard);
+            }
+            catch (OperationCanceledException)
+            {
```

Comment:
when we can get this?

Reply:
right, it shouldn't be there - done

Reply:
right, it shouldn't be there, done in https://github.com/ravendb/ravendb/pull/15922